src/client/js/client.js

```diff
@@ -1409,67 +1303,211 @@ define([
-                        //something funny is going on
-                        if (state.loadNodes[ROOT_PATH]) {
-                            //probably we are in the loading process,
-                            // so we should redo this update when the loading finishes
-                            //setTimeout(updateTerritory, 100, guid, patterns);
-                        } else {
-                            //root is not in nodes and has not even started to load it yet...
-                            state.users[guid].PATTERNS = JSON.parse(JSON.stringify(patterns));
+                        if (canSwitchStates()) {
+                            switchStates();
                         }
                     }
+                };
+
+            if (!state.nodes[ROOT_PATH]) {
+                if (state.users[guid]) {
+                    state.users[guid].PATTERNS = COPY(patterns);
+                }
+                return;
+            }
+
+            //empty territory check
+            if (keys.length === 0) {
+                _updateTerritoryAllDone(guid, patterns, null);
+                return;
+            }
+
+            state.ongoingTerritoryUpdateCounter += 1;
+
+            //first we have to set the internal counter as the actual load can get synchronous :(
+            loadRequestCounter = keys.length;
+
+
+            for (i = 0; i < keys.length; i += 1) {
+                if (state.inLoading) {
+                    state.ongoingLoadPatternsCounter += 1;
+                    loadPatternThrottled(state.core,
+                        keys[i], patterns[keys[i]], state.loadNodes, loadingPatternFinished);
                 } else {
-                    //we should update the patterns, but that is all
-                    state.users[guid].PATTERNS = JSON.parse(JSON.stringify(patterns));
+                    //we should save the patterns to a pending directory
+                    state.pendingTerritoryUpdatePatterns[updateRequestId] = patterns;
                 }
+                loadPatternThrottled(state.core, keys[i], patterns[keys[i]], state.nodes, patternLoaded);
             }
+
         };
 
+        function refreshMetaNodes(oldSource, newSource) {
+            var pathsToRemove = [],
+                i,
+                oldPaths = Object.keys(oldSource),
+                newPaths = Object.keys(newSource);
+
+            for (i = 0; i < oldPaths.length; i += 1) {
+                if (newPaths.indexOf(oldPaths[i]) === -1) {
+                    pathsToRemove.push(oldPaths[i]);
+                }
+            }
+
+            for (i = 0; i < newPaths.length; i += 1) {
+                state.metaNodes[newPaths[i]] = newSource[newPaths[i]].node;
+            }
+
+            for (i = 0; i < pathsToRemove.length; i += 1) {
+                delete state.metaNodes[pathsToRemove[i]];
+            }
+        }
+
+        function switchStates() {
+            //it is safe now to move the loadNodes into nodes,

```

Comment:
I would add more logging in general, but at least a debug message that we're switching state here.